features/mbedtls/targets/TARGET_STM/aes_alt.c

```diff
@@ -43,125 +48,147 @@
-            SWAP_B8_TO_B32(ctx->aes_key[2], key, 8);
-            SWAP_B8_TO_B32(ctx->aes_key[3], key, 12);
-            SWAP_B8_TO_B32(ctx->aes_key[4], key, 16);
-            SWAP_B8_TO_B32(ctx->aes_key[5], key, 20);
-            SWAP_B8_TO_B32(ctx->aes_key[6], key, 24);
-            SWAP_B8_TO_B32(ctx->aes_key[7], key, 28);
-
             break;
+
         default :
-            return (MBEDTLS_ERR_AES_INVALID_KEY_LENGTH);
+            ret = MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+            goto exit;
+    }
+
+    /* Format and fill AES key  */
+    for (i = 0; i < (keybits / 32); i++) {
+        GET_UINT32_BE(ctx->aes_key[i], key, 4 * i);
     }
 
     ctx->hcryp_aes.Init.DataType = CRYP_DATATYPE_8B;
-    ctx->hcryp_aes.Instance = CRYP;
+    ctx->hcryp_aes.Init.DataWidthUnit = CRYP_DATAWIDTHUNIT_BYTE;
     ctx->hcryp_aes.Init.pKey = ctx->aes_key;
 
-    /* Enable CRYP clock */
-    __HAL_RCC_CRYP_CLK_ENABLE();
-
-    if (HAL_CRYP_Init(&ctx->hcryp_aes) == HAL_ERROR) {
-        return (HAL_ERROR);
+    if (HAL_CRYP_Init(&ctx->hcryp_aes) != HAL_OK) {
+        ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+        goto exit;
     }
 
-    /* allow multi-instance of CRYP use: save context for CRYP HW module CR */
+    /* allow multi-context of CRYP : save context */
     ctx->ctx_save_cr = ctx->hcryp_aes.Instance->CR;
-    return (0);
 
-}
-
-/* Implementation that should never be optimized out by the compiler */
-static void mbedtls_zeroize(void *v, size_t n)
-{
-    volatile unsigned char *p = (unsigned char *)v;
-    while (n--) {
-        *p++ = 0;
+exit :
+    /* Free context access */
+#if defined(MBEDTLS_THREADING_C)
+    if (mbedtls_mutex_unlock(&cryp_mutex) != 0) {
+        ret = MBEDTLS_ERR_THREADING_MUTEX_ERROR;
     }
+#endif /* MBEDTLS_THREADING_C */
+
+    return (ret);
 }
 
 void mbedtls_aes_init(mbedtls_aes_context *ctx)
 {
     AES_VALIDATE(ctx != NULL);
 
-    memset(ctx, 0, sizeof(mbedtls_aes_context));
+    __disable_irq();
+#if defined(MBEDTLS_THREADING_C)
+    /* mutex cannot be initialized twice */
+    if (!cryp_mutex_started) {
+        mbedtls_mutex_init(&cryp_mutex);
+        cryp_mutex_started = 1;
+    }
+#endif /* MBEDTLS_THREADING_C */
+    cryp_context_count++;
+    __enable_irq();
+
+    cryp_zeroize((void *)ctx, sizeof(mbedtls_aes_context));
+
+#if (MBEDTLS_DEBUG)
+    printf("[ALT] mbedtls_aes_init %u\n", cryp_context_count);
+#endif
 }
 
 
 void mbedtls_aes_free(mbedtls_aes_context *ctx)
 {
+#if (MBEDTLS_DEBUG)
+    printf("[ALT] mbedtls_aes_free %u\n", cryp_context_count);
+#endif
 
     if (ctx == NULL) {
         return;
     }
-    /* Force the CRYP Periheral Clock Reset */
-    __HAL_RCC_CRYP_FORCE_RESET();
 
-    /* Release the CRYP Periheral Clock Reset */
-    __HAL_RCC_CRYP_RELEASE_RESET();
+    __disable_irq();
+    if (cryp_context_count > 0) {
+        cryp_context_count--;
+    }
 
-    mbedtls_zeroize(ctx, sizeof(mbedtls_aes_context));
+#if defined(MBEDTLS_THREADING_C)
+    if (cryp_mutex_started) {
```

Comment:
Only if `cryp_context_count` is now zero, surely? But as with the `mbedtls_mutex_init`, this wouldn't be legal with interrupts disabled.

Reply:
@tbistm  as pointed by @kjbracey-arm -  shouldn't you check that cryp_context_count is 0 ?

Reply:
As pointed by @kjbracey-arm, mutex should be freed when it is no more used. This point is relevant and valid.